### PR TITLE
Add guide for vscode with slurm

### PR DESCRIPTION
**Pull request by lauraporta**
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
We've been exploring optimal ways to use VSCode within a SLURM job to ensure appropriate use of assigned resources. After testing a solution suggested in our internal Slack, `code tunnel`, I found it to be effective. This approach is straightforward, user-friendly, and fully SLURM-managed, allowing resources to be properly allocated and controlled within the job environment.

Happy to hear your thoughts about it!

**What does this PR do?**
Adds a markdown file with the guide.

## References
#62 issue discussion

## How has this PR been tested?
- Docs test: docs built locally 
- Resource management tests: 
	- I can successfully kill the job and be disconnected from vscode

## Checklist:

- [x] The code has been tested locally
- [ ] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [ ] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by lauraporta** [comment]
Working on failing actions

File: docs/source/programming/vscode-with-slurm-job.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ssh username@swc.ucl.ac.uk
+```
+
+Once connected, request an interactive job via SLURM to access a compute node:
+
+```console
+srun -p fast --pty bash -i
+```
+
+In this example, `-p fast` requests the fast partition, with default time settings, though you may adjust this according to your needs.
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:22]: I meant it as a basic example, but yes I can explicitly ask for more resources

File: docs/source/programming/vscode-with-slurm-job.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+After connecting to a compute node, initiate VSCode Code Tunnel by typing:
+
+```console
+code tunnel
+```
+
+A URL will appear in the terminal, `https://github.com/login/device`. Follow this link, log in with your GitHub credentials, and enter the provided PIN to authorize access. After completing this step, a second link will appear in the terminal (e.g., `https://vscode.dev/tunnel/gpu-XXXX`), which you can follow to access your VSCode session running directly on the HPC compute node.
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:30]: From [vscode docs](https://code.visualstudio.com/docs/remote/tunnels#:~:text=When%20opening%20a%20vscode.,right%20set%20of%20remote%20machines.):
> Tunneling securely transmits data from one network to another via [Microsoft dev tunnels](https://learn.microsoft.com/azure/developer/dev-tunnels/overview).
>
> Both hosting and connecting to a tunnel requires authentication with the same Github or Microsoft account on each end. In both cases, VS Code will make outbound connections to a service hosted in Azure; no firewall changes are generally necessary, and VS Code doesn't set up any network listeners.
>
>Once you connect from a remote VS Code instance, an SSH connection is created over the tunnel in order to provide end-to-end encryption.

File: docs/source/programming/vscode-with-slurm-job.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+When you’re finished, simply exit the SLURM session to close the VSCode tunnel and release resources.
+
+### Important Note on Extensions
+
+When using VSCode in this setup, exercise caution with extensions. Some, like the Typescript and Javascript extension, are not recommended by IT due to potential performance issues. Install only essential extensions to maintain system efficiency.
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:36]: Actually, maybe not. The extensions will be stored in `~/.vscode/extensions` in our `/nfs/nhome/live/username` home folder... But if one opens the tunnel from a node and not from the gateway node it should be OK.

File: docs/source/programming/vscode-with-slurm-job.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamltyson) [reply docs/source/programming/vscode-with-slurm-job.md:31]: Maybe link to this information in the guide? I think a lot of users will be suspicious that they have to enter login details for this.

File: docs/source/programming/vscode-with-slurm-job.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamltyson) [reply docs/source/programming/vscode-with-slurm-job.md:37]: Yeah presumably the VSCode process is running within SLURM? In that case it doesn't matter how many extensions the user has.

[the remark this answers is not in the document]
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:37]: Agreed

**Review by niksirbi — APPROVED** [review]
Nice find @lauraporta!

As far as I can tell this works as intended, and is the most hussle-free method I've tried so far and the only pre-requisite is a github account.

I could also login with my VScode account, sync all my extensions, settings etc, all working seamlessly!

I also ran some jupyter notebook within VSCode, and everything worked perfectly well, including plots.


SLURM killed my jobs at the specified time limit, so we shouldn't be eating more resources than allocated.

All in all, 10/10!

It my be worth announcing this on the institute's slack as the "recommended" way to use VSCode on the cluster (after consulting with the scientific computing team perhaps).

**Comment by sfmig** [comment]
Heya,
Nice find! ✨ 

I was just having a go now after the positive reviews but unfortunately couldn't make it work for me - it stays forever in the VSCode webapp saying that it is connecting to host.

I followed the instructions in the guide, except that I requested a node with a gpu in the `gpu` partition (rather than in `fast`):
```
srun -p gpu --gres=gpu:1 --pty bash -i
```
I seemed to get into the compute node just fine, then all the authentication went ok and the webapp launched, but it didn't move pass that.

I tried a few times with no success, and later saw that the SLURM jobs appear as FAILED:
image

I wonder if other people experienced this?

**Comment by sfmig** [comment]
And just two small comments on the guide:
- if you run `code tunnel` from the VSCode terminal, you get an error. Maybe we should clarify one should run this in a separate one?
- From the docs it seems that it [eliminates the need to copy local code in the compute node](https://code.visualstudio.com/docs/remote/tunnels#:~:text=This%20can%20eliminate%20the%20need%20for%20source%20code%20to%20be%20on%20your%20VS%20Code%20client%20machine%20since%20the%20extension%20runs%20commands%20and%20other%20extensions%20directly%20on%20the%20remote%20machine.). I haven't been able to check this, but is that correct? if so, should we highlight this as the main benefit? I didn't get that from reading the guide alone and it seems like its main perk.

**Comment by niksirbi** [comment]
> I was just having a go now after the positive reviews but unfortunately couldn't make it work for me - it stays forever in the VSCode webapp saying that it is connecting to host.

Strange. I wonder if the resources given to the compute node "by default" are not sufficient for whatever VSCode dose when tunneling. What happens if you try:
```bash
srun -p gpu -n 4 --mem 8G --gres=gpu:1 --pty bash -i
```

Just a random guess, not sure if this is indeed your problem.

**Comment by sfmig** [comment]
If you want to use your own VScode app rather than the webapp, it seems you can do this (from this [blog post](https://dev.to/burkeholland/introducing-vs-code-remote-tunnels-connect-to-remote-machines-with-ease-3nlg)):
1 - ssh to hpc-gw1
2- request an interactive compute node
3- run `code tunnel` in the compute node
4- in your vscode app, open the command palette (in mac, cmd+shift+p) and search "Remote tunnels: connect to tunnel". Then a dropdown shows up with the existing tunnels and you select the desired one.

I haven't been able to verify this 100% because it also lags forever (but at least it behaves the same 😅 )

**Comment by niksirbi** [comment]
> If you want to use your own VScode app rather than the webapp, it seems you can do this (from this [blog post](https://dev.to/burkeholland/introducing-vs-code-remote-tunnels-connect-to-remote-machines-with-ease-3nlg)):

It works! Nice find @sfmig, this is even better than suing the web app!

**Comment by sfmig** [comment]
> It works! 
yay!

I can confirm the steps above now also work for me with the two `srun` commands we discussed

**Comment by lauraporta** [comment]
>if you run code tunnel from the VSCode terminal, you get an error. Maybe we should clarify one should run this in a separate terminal?

Definitely yes. 


>From the docs it seems that it [eliminates the need to copy local code in the compute node](https://code.visualstudio.com/docs/remote/tunnels#:~:text=This%20can%20eliminate%20the%20need%20for%20source%20code%20to%20be%20on%20your%20VS%20Code%20client%20machine%20since%20the%20extension%20runs%20commands%20and%20other%20extensions%20directly%20on%20the%20remote%20machine.). I haven't been able to check this, but is that correct? if so, should we highlight this as the main benefit? I didn't get that from reading the guide alone and it seems like its main perk.

Interesting. Reading the highlighted section of the guide you point to, I think it means that we are not running source code on our local client (in this case the browser) but instead on the compute node in which code tunnel is run.

> If you want to use your own VScode app rather than the webapp...

Going to test this as well!

**Comment by lauraporta** [comment]
✨ A M A Z I N G ✨

I just didn't have the Tunnel extension installed. As the tunnel is maintained by the same kind of slurm job, it should be killed in the same way. Going to include this solution.

Anyway is anyone getting the notification "Unable to watch for changes"? 
Maybe it is just my repo.

File: docs/source/programming/vscode-with-slurm-job.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# Using VSCode with SLURM-Managed jobs on the SWC HPC cluster
+
+This guide explains how to set up and use VSCode within a SLURM-managed job on the SWC HPC cluster, offering a solution for users who require fast access to shared storage or substantial computational resources.
+
+This solution is easy to set up and constrained by the SLURM job's resource limits, ensuring that users can work efficiently without overloading the system.
+
+
+## Instructions
+
+First, connect to the bastion node by running:
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:10]: Ah alright I was reading the sentence in another way. First is the bastion and that's what I meant. But it makes more sense to talk about the gateway, as it's the last one we connect to.

**Comment by sfmig** [comment]
> I think it means that we are not running source code on our local client (in this case the browser) but instead on the compute node in which code tunnel is run.

Aaah I see! Sorry I wasn't sure who the client was 😅

**Comment by sfmig** [comment]
A somewhat equivalent approach using `ssh` would be the well-known hack:
1. ssh to `hpc-gw1` using VScode Remote extension 
    - (this is just to get the same UI experience as the tunnel)
1. request an interactive node with `srun`
	- suppose we get `gpu-380-11`
1. ssh to the assigned node by adding a SSH remote connection in VScode and typing : 
    `ssh -J hpc-gw1 sminano@gpu-380-11`
    - use `-J` to indicate a proxy jump via the gateway node `hpc-gw1`
	- the node gets added as a host to your local config 🙀 

This works for me at the moment. 

The problem of this approach is that the SLURM limits assigned to the user in the `srun` command are not maintained when you `ssh` to the compute node. Is that correct?
	
The benefit of the `code tunnel` approach would then be:
- it's easier to set up, and
- we think it respects the SLURM limits (or do we know this for sure?)

File: docs/source/programming/vscode-with-slurm-job.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+After connecting to a compute node, initiate VSCode Code Tunnel by typing:
+
+```bash
+code tunnel
+```
+
+A URL will appear in the terminal, `https://github.com/login/device`. 
+Follow this link, log in with your GitHub credentials, and enter the provided PIN to authorize access. 
+After completing this step, a second link will appear in the terminal (e.g., `https://vscode.dev/tunnel/gpu-XXXX`), which you can follow to access your VSCode session running directly on the HPC compute node.
+NB: the GPU name in the URL provided might not match the actual node name.
```
Comment (sfmig) [inline docs/source/programming/vscode-with-slurm-job.md:34]: I was confused by this btw, thanks for adding the clarification 🚀

**Comment by niksirbi** [comment]
> A somewhat equivalent approach using ssh would be the well-known hack.

This is what I had used a few times before, and I have discussed with a few others who also used it. The gist is, we are not sure  if this would respect SLURM constraints. The problem is that while you are requesting the node with SLURM, then you directly `ssh` into the assigned node and any jobs you run inherit from that `ssh` jump, not from the `srun`. It sounds like this could work, if some plugins were configured in a certain way (by the IT), but they probably are not. In essense, if you `ssh` into a node, you can run whatever you want in that node without constraints (which you shouldn't), unless IT somehow configures it to "be aware" of your constraints based on your user name and the previously run `srun` command.

The `code tunnel` approach is different, because the launching of the tunnel itself happens within an interactive `srun` session, so it *should* inherit the constraints imposed by `srun`.

To summarise, we don't know for sure without further tests, but the `code tunnel` appoach is definitely more likely to respect SLURM (and in the case of the time limit, it demonstrably does). If we follow you hack @sfmig, but impose say a 5min limit on `srun`, then `ssh`, I'd bet that your remote ssh connection to the compute node will not be killed when the 5 mins are up (happy to be proven wrong about this!).

**Comment by sfmig** [comment]
Yes agree with all, I was aware with the issues of the previous ssh approach, was just trying to summarise and confirm what we have evidence for 😅. 

But yeah worth highlighting the bit on why code tunnel may have more chances of respecting the slurm limits.

**Comment by niksirbi** [comment]
> (also maybe don't call it _my_ hack, I had to fish it from old conversations in the hpc channel and 100% don't endorse it 😂)

Yeah, sorry, I'm aware, that's when I also saw it to begin with 😉.
I was using "your" as a shorthand for "the one you mentioned above".

**Comment by niksirbi** [comment]
> Yep it did get killed, @lauraporta and I actually did this check at the time

NIce! Then I remain forever confused by this topic.

**Comment by sfmig** [comment]
sorry, I wrote the opposite of what I was thinking, it did *not* get killed, sorry to add to the confusion 🫠  - edited now

(and gonna go for lunch before I say more silly things)

**Comment by niksirbi** [comment]
> sorry, I wrote the opposite of what I was thinking, it did _not_ get killed, sorry to add to the confusion 🫠 - edited now

Phew, a bit of confidence in my own understanding of this restored.

**Comment by lauraporta** [comment]
Apart of SLURM (Scheduling Like Unpredictable Resource Magic) questions, what do you think of the updated guide?

File: docs/source/programming/vscode-with-slurm-job.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+```{code-block} console
+$ code tunnel
+```
+
+A URL will appear in the terminal, `https://github.com/login/device`.
+Follow this link, log in with your GitHub credentials, and enter the provided PIN to authorize access.
+
+You have two options to run VSCode:
+
+-  **Run VSCode in$  the browser**
```
Comment (niksirbi) [inline docs/source/programming/vscode-with-slurm-job.md:36]:
```suggestion
-  **Run VSCode in  the browser:**
```

File: docs/source/programming/vscode-with-slurm-job.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    After completing the above step, a second link will appear in the terminal (e.g., `https://vscode.dev/tunnel/gpu-XXXX`), which you can follow to access your VSCode session running directly on the HPC compute node.
+
+- **Run VSCode on your local machine**
```
Comment (niksirbi) [inline docs/source/programming/vscode-with-slurm-job.md:39]:
```suggestion
- **Run VSCode on your local machine:**
```

Line breaks are not properly rendered by Sphinx with the buller point. I sugeest adding a color (as in my suggestion) or a period, to signal separation between the bold part and the main text.

File: docs/source/programming/vscode-with-slurm-job.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+$ code tunnel
+```
+
+A URL will appear in the terminal, `https://github.com/login/device`.
+Follow this link, log in with your GitHub credentials, and enter the provided PIN to authorize access.
+
+You have two options to run VSCode:
+
+-  **Run VSCode in  the browser:**
+    After completing the above step, a second link will appear in the terminal (e.g., `https://vscode.dev/tunnel/gpu-XXXX`), which you can follow to access your VSCode session running directly on the HPC compute node.
```
Comment (niksirbi) [inline docs/source/programming/vscode-with-slurm-job.md:37]: note that not all node names start with `gpu` on our cluster. Some start with `enc` etc.
Should we just say `https://vscode.dev/tunnel/<node-name>` to catch all cases? And similarly for the othe mentions of `gpu-XXXX` later.

File: docs/source/programming/vscode-with-slurm-job.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+> Tunneling securely transmits data from one network to another via [Microsoft dev tunnels](https://learn.microsoft.com/azure/developer/dev-tunnels/overview).
+>
+> Both hosting and connecting to a tunnel requires authentication with the same Github or Microsoft account on each end. In both cases, VS Code will make outbound connections to a service hosted in Azure; no firewall changes are generally necessary, and VS Code doesn't set up any network listeners.
+>
+>Once you connect from a remote VS Code instance, an SSH connection is created over the tunnel in order to provide end-to-end encryption.
+:::
+
+## Additional benefits of code tunnel
+
+One advantage of using VSCode's code tunnel is that it forwards any HTTP servers launched from the same node, such as MLflow or Jupyter Notebook servers. When these are initiated, VSCode will notify you with a link that you can follow to access the server's UI directly.
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md]: It works for dash-plotly as well 🥰 🔥

File: docs/source/programming/vscode-with-slurm-job.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+You have two options to run VSCode:
+
+-  **Run VSCode in  the browser:**
+    After completing the above step, a second link will appear in the terminal (e.g., `https://vscode.dev/tunnel/gpu-XXXX`), which you can follow to access your VSCode session running directly on the HPC compute node.
+
+- **Run VSCode on your local machine:**
+    If you want instead to use your local VSCode, install the "Remote - Tunnels" extension, click on "Open remote window" in the bottom left corner of the VSCode window, and select "Connect to Tunnel". You should see the name "gpu-XXXX" in the list of available tunnels. Click on it to connect to the VSCode session running on the HPC compute node.
+
+NB: the GPU name in the URL provided might not match the actual node name.
```
Comment (niksirbi) [inline docs/source/programming/vscode-with-slurm-job.md:42]:
```suggestion
NB: the node name in the provided URL might not match the actual node name.
```

Let's stick with the "node name" terminology without specific reference to `gpu` (see comment above).

File: docs/source/programming/vscode-with-slurm-job.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+If by mistake you close your terminal window, the tunnel will continue to run till you reach the time limit. To rejoin the slurm job, you can use the following command if you know the job ID:
```
Comment (niksirbi) [inline docs/source/programming/vscode-with-slurm-job.md:44]: Nice!

File: docs/source/programming/vscode-with-slurm-job.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+>
+>Once you connect from a remote VS Code instance, an SSH connection is created over the tunnel in order to provide end-to-end encryption.
+:::
+
+## Additional benefits of code tunnel
+
+One advantage of using VSCode's code tunnel is that it forwards any HTTP servers launched from the same node, such as Dash-Plotly apps or Jupyter Notebook servers. To launch your additional server, request a separate slurm job for the same compute node, e.g.:
+
+```{code-block} console
+$ srun -p gpu -w <gpu-XXXX> -n 4 --mem 8G --pty bash -i
```
Comment (niksirbi) [inline docs/source/programming/vscode-with-slurm-job.md:69]: I like the example! It makes it clearer.

File: docs/source/programming/vscode-with-slurm-job.md (hunk 8 above)
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:37]: Like it

File: docs/source/programming/vscode-with-slurm-job.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+```
+When these are initiated, VSCode will notify you with a link that you can follow to access the server's UI directly.
+
+In order to do so, can request a separate slurm job for the same node to run your alternative sever
```
Comment (niksirbi) [inline docs/source/programming/vscode-with-slurm-job.md:73]: Looks like this sentence is duplicate and should be removed?

**Comment by niksirbi** [comment]
Thanks for updating it @lauraporta. I'm happy with the content, just added some comments on wording and rendering.

File: docs/source/programming/vscode-with-slurm-job.md (hunk 13 above)
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:73]: Ah yes good catch

**Review by sfmig — APPROVED** [review]
this was a great find, thanks @lauraporta 🚀 

Just small things such as some vscode ->VSCode and slurm -> SLURM. Nothing major, feel free to take or leave as you please!

File: docs/source/programming/vscode-with-slurm-job.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
+# Using VSCode with SLURM-managed jobs on the SWC HPC cluster
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply docs/source/programming/vscode-with-slurm-job.md:1]: how about "interactive SLUR jobs"?

File: docs/source/programming/vscode-with-slurm-job.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+This guide explains how to set up and use VSCode within a SLURM-managed job on the SWC HPC cluster, offering a solution for users who require fast access to shared storage or substantial computational resources.
+
+This solution is easy to set up and constrained by the SLURM job's resource limits, ensuring that users can work efficiently without overloading the system.
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:5]: Well I think the point of doing all this is that if SLURM is restrictive, so will be anything that we run with VSCode tunnel. If we remove this point, we loose the advantage of using code-tunnel instead of ssh-ing into a node.
I can make it more explicit in the sense that is constrained as long as SLURM constrains 😓

[the remark this answers is not in the document]
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:5]: Also, time limits are there, and my programs still get killed by slurm if I don't ask for enough memory. I am unsure why with numpy array they didn't get killed.

File: docs/source/programming/vscode-with-slurm-job.md [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+::: {dropdown} Why do I have to authenticate via GitHub?
+:color: info
+:icon: info
+
+As explained in [vscode docs](https://code.visualstudio.com/docs/remote/tunnels) it serves as a secure way to authenticate the user and ensure that only the user who initiated the tunnel can access it:
+> Tunneling securely transmits data from one network to another via [Microsoft dev tunnels](https://learn.microsoft.com/azure/developer/dev-tunnels/overview).
+>
+> Both hosting and connecting to a tunnel requires authentication with the same Github or Microsoft account on each end. In both cases, VS Code will make outbound connections to a service hosted in Azure; no firewall changes are generally necessary, and VS Code doesn't set up any network listeners.
+>
+>Once you connect from a remote VS Code instance, an SSH connection is created over the tunnel in order to provide end-to-end encryption.
```
[the remark this answers is not in the document]
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:61]: 😭

[the remark this answers is not in the document]
Comment (lauraporta) [reply docs/source/programming/vscode-with-slurm-job.md:63]: In the rest of the guides we are using VSCode, so... well... I'll leave it like that

File: docs/source/programming/vscode-with-slurm-job.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (sfmig) [reply docs/source/programming/vscode-with-slurm-job.md]: aah ok I see! thanks for checking this
